File: leetcode/go/367.valid-perfect-square.go
func isPerfectSquare(num int) bool {
    sqrtn := 1
    err := 1
    for err > 0 {
        sqrtn = (sqrtn + num/sqrtn)/2
        err = sqrtn - num/sqrtn
    }
    return sqrtn * sqrtn == num
}

File: leetcode/rust-leetcode/src/367.valid-perfect-square.rs
impl Solution {
    pub fn is_perfect_square(num: i32) -> bool {
        let mut i: i32 = 1;
        let mut e: i32 = 1;
        while i - num / i > 0 {
            i = (i + num / i)/2;
            e = i - num / i;
        }
        i * i == num
    }
}